File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/library/items/media/audio/uri/specs/GivenACachedFile/WhenProvidingTheUri.java
package com.lasthopesoftware.bluewater.client.browsing.library.items.media.audio.uri.specs.GivenACachedFile;

import android.net.Uri;

import com.lasthopesoftware.bluewater.client.browsing.library.items.media.audio.uri.CachedAudioFileUriProvider;
import com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.cached.access.ICachedFilesProvider;
import com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.cached.repository.CachedFile;
import com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.uri.RemoteFileUriProvider;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.io.File;
import java.io.IOException;

import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class WhenProvidingTheUri {

	private static Uri cachedFileUri;

	private static final File file;

	static {
		File file1;
		try {
			file1 = File.createTempFile("temp", ".txt");
		} catch (IOException e) {
			e.printStackTrace();
			file1 = new File("test");
		}

		file = file1;
		file.deleteOnExit();
	}

	@BeforeClass
	public static void before() {
		final Uri remoteUri = Uri.parse("http://a-url/file?key=1");
		final RemoteFileUriProvider remoteFileUriProvider = mock(RemoteFileUriProvider.class);
		when(remoteFileUriProvider.promiseFileUri(new ServiceFile(10)))
			.thenReturn(new Promise<>(remoteUri));

		final ICachedFilesProvider cachedFilesProvider = mock(ICachedFilesProvider.class);
		when(cachedFilesProvider.promiseCachedFile(remoteUri.getPath() + "?" + remoteUri.getQuery()))
			.thenReturn(new Promise<>(new CachedFile().setFileName(file.getAbsolutePath())));

		final CachedAudioFileUriProvider cachedAudioFileUriProvider =
			new CachedAudioFileUriProvider(
				remoteFileUriProvider,
				cachedFilesProvider);

		cachedAudioFileUriProvider
			.promiseFileUri(new ServiceFile(10))
			.then(uri -> cachedFileUri = uri);
	}

	@Test
	public void thenTheUriIsThePathToTheFile() {
		assertThat(cachedFileUri.toString()).isEqualTo(Uri.fromFile(file).toString());
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/browsing/items/media/audio/uri/GivenACachedFile/WhenProvidingTheUri.kt
package com.lasthopesoftware.bluewater.client.browsing.items.media.audio.uri.GivenACachedFile

import android.net.Uri
import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.files.cached.CacheFiles
import com.lasthopesoftware.bluewater.client.browsing.files.uri.RemoteFileUriProvider
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.playback.caching.uri.CachedAudioFileUriProvider
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.io.File

private const val libraryId = 408

@RunWith(RobolectricTestRunner::class)
class WhenProvidingTheUri {
	companion object {
		private val file by lazy {
			File.createTempFile("temp", ".txt").apply { deleteOnExit() }
		}

		private val cachedFileUri by lazy {
			val remoteUri = Uri.parse("http://a-url/file?key=1")
			val remoteFileUriProvider = mockk<RemoteFileUriProvider>().apply {
				every { promiseUri(LibraryId(libraryId), ServiceFile("10")) } returns Promise(remoteUri)
			}

			val cachedFilesProvider = mockk<CacheFiles> {
				every { promiseCachedFile(LibraryId(libraryId), remoteUri.path + "?" + remoteUri.query) } returns Promise(file)
			}

			val cachedAudioFileUriProvider = CachedAudioFileUriProvider(
				remoteFileUriProvider,
				cachedFilesProvider
			)

			cachedAudioFileUriProvider
				.promiseUri(LibraryId(libraryId), ServiceFile("10"))
				.toExpiringFuture()
				.get()
		}
	}

	@Test
	fun thenTheUriIsThePathToTheFile() {
		assertThat(cachedFileUri.toString())
			.isEqualTo(Uri.fromFile(file).toString())
	}
}
